### PR TITLE
Implement API Create Offer

### PR DESCRIPTION
**Pull request by dplumb94**

### CODE REVIEW

File: transaction_creation/marketplace_transaction/transaction_creation.py [hunk 1]
```diff
@@ -143,3 +143,58 @@ def create_holding(txn_key,
...
+    inputs = [
+        addresser.make_account_address(
+            account_id=txn_key.get_public_key().as_hex()),
+        addresser.make_holding_address(holding_id=source),
+        addresser.make_holding_address(holding_id=target),
+        addresser.make_offer_address(offer_id=identifier)
+    ]
+
+    outputs = [addresser.make_offer_address(offer_id=identifier),
+               addresser.make_account_address(
```
Comment (boydjohnson) [inline transaction_creation/marketplace_transaction/transaction_creation.py]: I don't think the Account is an output of the txn. Only the Offer will be created, and no holdings will be modified until the Offer is Accepted.

**Review by boydjohnson — CHANGES_REQUESTED** [review]
Outputs should be slightly modified in create_offer.

**Review by boydjohnson — APPROVED** [review]

File: rest_api/api/offers.py [hunk 2]
```diff
@@ -53,3 +85,21 @@
...
+def _create_offer_dict(body, public_key):
+    keys = ['label', 'description', 'source', 'rules',
+            'sourceQuantity', 'target', 'targetQuantity']
+
+    offer = {k: body[k] for k in keys if body.get(k) is not None}
+
+    if offer.get('sourceQuantity') is None:
+        offer['sourceQuantity'] = 0
+    if offer.get('targetQuantity') is None:
+        offer['targetQuantity'] = 0
```
Comment (delventhalz) [inline rest_api/api/offers.py]: I'm not sure this correct given our recent Slack discussion. A source quantity of 0 is never valid, and a target quantity of 0 is only valid if target is also unspecified. I think it is probably best to leave these fields unset. In most cases, the validator will spit back an invalid transaction anyway. In the event that both target and targetQuantity are unset, leaving them unset in the response object seems sensible to me.

File: rest_api/api-spec.yaml [hunk 3]
```diff
@@ -696,6 +696,8 @@ definitions:
       A simple enum and with optional value, used to constrain and define the
       behavior of Assets or Offers
     type: object
+    required:
+      - type
```
Comment (delventhalz) [inline rest_api/api-spec.yaml:700]: As per our recent Slack discussion, `target` and `targetQuantity` should be removed from the `required` field of `NewOfferBody`

File: rest_api/api/offers.py [hunk 4]
```diff
@@ -26,7 +33,32 @@
 @authorized()
 async def create_offer(request):
     """Creates a new Offer in state"""
-    raise ApiNotImplemented()
+    required_fields = ['source', 'sourceQuantity', 'target', 'targetQuantity']
```
Comment (delventhalz) [inline rest_api/api/offers.py]: `target` and `targetQuantity` are not required.

**Review by delventhalz — CHANGES_REQUESTED** [review]

**Comment by delventhalz** [comment]
Also, a hook needs to be added so this test passes:
```
fail: POST (200) /offers duration: 271ms
fail: statusCode: Status code is not '200'
body: 
{
  "error": "Bad Request: Failed to create Offer, Holding id eiusmod listed as source does not refer to a Holding."
}
statusCode: 400
```

**Review by delventhalz — APPROVED** [review]

**Pull request merged**